### PR TITLE
Nio read attr

### PR DESCRIPTION
**Pull request by rickle-msft**

### CODE REVIEW

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    private final BlobProperties properties;
+
+    /*
+    There are some work-arounds we could do to try to accommodate virtual directories such as making a checkDirStatus
+    call before or after getProperties to throw an appropriate error or adding an isVirtualDirectory method. However,
+    the former wastes network time only to throw a slightly more specific error when we will throw on 404 anyway. The
+    latter introduces virtual directories into the actual code path/api surface. While we are clear in our docs about
+    the possible pitfalls of virtual directories, and customers should be aware of it, they shouldn't have to code
+    against it. Therefore, we fall back to documenting that reading attributes on a virtual directory will throw. And
+    who reads attributes on a directory anyway?
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:55]: nit: May want to remove the last line 😃

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    public boolean isDirectory() {
+        return !this.isRegularFile();
+    }
+
+    /**
+     * @return false. Symbolic links are not supported.
+     */
+    @Override
+    public boolean isSymbolicLink() {
+        return false;
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:115]: Should this throw `UnsupportedOperationException` if it isn't supported?

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributeView.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                throw LoggingUtility.logError(view.logger, new UncheckedIOException(ATTR_CONSUMER_ERROR, e));
+            }
+        });
+        map.put("metadata", obj -> {
+            try {
+                Map<String, String> m = (Map<String, String>) obj;
+                if (m == null) {
+                    throw LoggingUtility.logError(view.logger, new ClassCastException());
+                }
+                view.setMetadata((Map<String, String>) obj);
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributeView.java:55]: Should be able to pass `m` here.

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributes.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+     * {@link AzureFileSystemProvider#createDirectory(Path, FileAttribute[])} for more information on virtual and
+     * concrete directories.
+     */
+    @Override
+    public boolean isDirectory() {
+        return !this.isRegularFile();
+    }
+
+    /**
+     * @return false. Symbolic links are not supported.
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributes.java:259]: Same question about throwing `UnsupportedOperationException`.

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java (hunk 1 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:55]: Yeaa probably :(

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java (hunk 2 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:115]: I'm not sure. In some sense it's just always false, right? But maybe it makes it more clear that sym links aren't supported if I throw?

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributeView.java (hunk 3 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributeView.java:55]: Good catch

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java (hunk 2 above)
Comment (kasobol-msft) [reply sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:115]: I think returning always `false` is correct. Throwing `UnsupportedOperationException` would mean "I cannot check if file is symlink or not` in plain English. I believe our file system doesn't have symlinks so all files are real by definition, therefore we know how to check if they're symlinks or not.

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    call before or after getProperties to throw an appropriate error or adding an isVirtualDirectory method. However,
+    the former wastes network time only to throw a slightly more specific error when we will throw on 404 anyway. The
+    latter introduces virtual directories into the actual code path/api surface. While we are clear in our docs about
+    the possible pitfalls of virtual directories, and customers should be aware of it, they shouldn't have to code
+    against it. Therefore, we fall back to documenting that reading attributes on a virtual directory will throw. And
+    who reads attributes on a directory anyway?
+     */
+    AzureBasicFileAttributes(Path path) throws IOException {
+        try {
+            this.properties = new AzureResource(path).getBlobClient().getProperties();
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:59]: What if properties change while someone is holding reference to this object? I.e. lastModifiedTime shifts between calling this constructor and lastModifiedTime()? 
Could you check what happens if you use local file system?

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        return FileTime.from(properties.getLastModified().toInstant());
+    }
+
+    /**
+     * Unsupported.
+     * {@inheritDoc}
+     */
+    @Override
+    public FileTime lastAccessTime() {
+        throw new UnsupportedOperationException();
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:79]: Might want to add @throws to javadocs for unsupported apis.

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributes.java [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        - encryption key sha256
+     */
+
+    private final ClientLogger logger = new ClientLogger(AzureBlobFileAttributes.class);
+
+    private final BlobProperties properties;
+
+    AzureBlobFileAttributes(Path path) throws IOException {
+        try {
+            this.properties = new AzureResource(path).getBlobClient().getProperties();
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributes.java:47]: same question here about keeping state.

File: sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java (hunk 5 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:59]: It looks like the attributes always return the same value once you've read them, and if you want an updated value, you need to make another call to read attributes, which is what we do here

Comment (kasobol-msft) [reply sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBasicFileAttributes.java:58]: cool, thanks for checking.
